### ISSUE TITLE
Segmentation fault 11 / Python on Mac

### ISSUE THREAD

**Issue by ruslanolkhovsky** [jruizgit/rules#306]
Hello,

I experience segmentation fault 11 in python on Mac. Here is part of the trace:

`
Current thread 0x0000700000c10000 (most recent call first):
  File "/Users/ruslan/.virtualenvs/platform/lib/python3.7/site-packages/durable/engine.py", line 429 in dispatch_timers
  File "/Users/ruslan/.virtualenvs/platform/lib/python3.7/site-packages/durable/engine.py", line 919 in dispatch_timers
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 1166 in run
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 885 in _bootstrap
`

Looks like there is an issue with the dispatch_timers, but I am not sure.

I have created a simple test (please, see the two code snippets below) to call the engine with events that results in a MemoryError. Please, check if you can reproduce. I appreciate your advice. Thank you!

**test.py** 

```
import sys
import os
import json
import time
import uuid
import threading
import psutil
import faulthandler

from durable.lang import *


m = {
    "timestamp": None,
    "sid": None,
    "command": 1,
    "id": 1,
    "type": 1,
    "payload": "500"
}


def load_from_file(name):
    with open(name, "r") as file:
        data = file.read()
    obj = json.loads(str(data))
    file.close()

    return obj


def memory_usage(i=None):
    process = psutil.Process(os.getpid())
    if i:
        mem = psutil.virtual_memory()[i]
    else:
        mem = psutil.virtual_memory()

    return mem


# Action methods
def action_first(c):
    print("Memory usage: {}%    Threads: {}".format(
        memory_usage(2), str(threading.active_count())))

    print("{}: run action_first, ".format(c.m.timestamp), end="")
    return


def action_second(c):
    print("action_second.")
    return


class Host(engine.Host):

    def __init__(self):
        super(Host, self).__init__()

    def get_action(self, action_name):

        def print_lambda(c):

            c.s.label = action_name
            run = getattr(sys.modules[__name__], action_name)
            run(c)

        return print_lambda


def main():
    h = Host()

    ruleset = load_from_file("ruleset.json")

    h.set_rulesets(ruleset)

    while True:
        m["timestamp"] = int(time.time())
        m["sid"] = str(uuid.uuid4())

        while True:
            r = h.post("example", m)
            if not isinstance(r, dict) or not r["running"]:
                break

        time.sleep(0.01)


if __name__ == "__main__":
    faulthandler.enable()
    main()


```
**ruleset.json**

```
{
  "example$flow": {
    "input": {
      "to": {
        "first": {
          "all": [
            {
              "m": {
                "command": 1
              }
            }
          ]
        }
      }
    },
    "first": {
      "run": "action_first",
      "to": {
        "second": {
          "all": [
            {
              "m": {
                "$and": [
                  {
                    "command": 1
                  },
                  {
                    "$neq": {
                      "id": 0
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    },
    "second": {
      "run": "action_second",
      "to": "end"
    },
    "end": {}
  }
}

```

**Comment by ruslanolkhovsky** [comment]
UPD. The same problem exists on the Debian GNU/Linux 10.

**Comment by GergelyMolnar** [comment]
Just from interest I ran your sample, but I couldn't reproduce the error on Centos 7.7 using Python 3 and durable-rules 2.0.19

```
$ uname -a
Linux ox0303-lt 5.3.1-1.el7.elrepo.x86_64 #1 SMP Sat Sep 21 09:44:09 EDT 2019 x86_64 x86_64 x86_64 GNU/Linux

$ python3 --version
Python 3.6.8

# running in a virtualenv with
$pip3 freeze
attrs==19.3.0
certifi==2019.9.11
cffi==1.13.2
chardet==3.0.4
coverage==5.0.3
cryptography==2.8
durable-rules==2.0.19
idna==2.8
importlib-metadata==1.5.0
more-itertools==8.2.0
packaging==20.1
pluggy==0.13.1
psutil==5.6.7
py==1.8.1
pycparser==2.19
PyJWT==1.7.1
pyparsing==2.4.6
pytest==5.3.5
pytest-cov==2.8.1
PyYAML==5.2
requests==2.22.0
six==1.13.0
urllib3==1.25.7
wcwidth==0.1.8
zipp==2.1.0
```

Instead, it stuck in an infinite loop of
```
Memory usage: 50.0%    Threads: 3
1581433713: run action_first, action_second.
Memory usage: 50.0%    Threads: 3
1581433713: run action_first, action_second.
Memory usage: 50.0%    Threads: 3
1581433713: run action_first, action_second.
Memory usage: 50.0%    Threads: 3
...
```

**Comment by ruslanolkhovsky** [comment]
@GergelyMolnar I did not try it on the Centos, used Mac OS and Debian. How long did you wait?
It is a loop as you can see from the code :) The memory is just for info. The timestamp should change though.

**Comment by ruslanolkhovsky** [comment]
I checked it in the Ubuntu 18.04.3 LTS, same problem: MemoryError in approx. 20 sec... Here is the trace:

```
Traceback (most recent call last):
  File "test.py", line 93, in <module>
    main()
  File "test.py", line 84, in main
    r = h.post("example", m)
  File "/home/edgica/test/venv/lib/python3.6/site-packages/durable/engine.py", line 786, in post
    return self._handle_function(rules, rules.assert_event, message, complete)
  File "/home/edgica/test/venv/lib/python3.6/site-packages/durable/engine.py", line 770, in _handle_function
    rules.do_actions(func(args), callback)
  File "/home/edgica/test/venv/lib/python3.6/site-packages/durable/engine.py", line 329, in assert_event
    return self._handle_result(durable_rules_engine.assert_event(self._handle, json.dumps(message, ensure_ascii=False)), message)
MemoryError

```

**Comment by ruslanolkhovsky** [comment]
Is there any chance to get this fixed? We are developing a system with the rules as a core element, so it is a serious blocker for us. Appreciate your help. Thank you.

**Comment by jruizgit** [comment]
Hi, I will be looking into it this week.

**Comment by jruizgit** [comment]
Hi, thanks for reporting this. I have debugged the scenario. Couple of things:

1. The test case is creating state objects without deleting them. I recommend adding a cleanup action at the end of the flow.

    In your json file, add the `action_cleanup`:
```javascript
    "end": {
      "run": "action_cleanup" 
    }
````
    In your python code, delete the state in the cleanup action:
```python
def action_cleanup(c):
    print("action_cleaup.")
    c.delete_state()
    return
```

2. There was a memory leak in the cleanup state code, for which I have published a fix. Please use version 2.0.21.

With the memory leak fixed, the process ran stable with 32.8% mem usage and 14.4MB (process monitor).

**Comment by ruslanolkhovsky** [comment]
Hi Jesus. Thank you for the support! It looks fine now at the first sight. I will test more and confirm close.

BTW, I noticed that you do not pass the 'm' variable to the 'c' parameter in action method if there were no antecedent filter condition for '$m' indicated. Is it for purpose? How would you recommend declaring the filter condition to pass 'm' in all cases?

Thank you.
